designs/2020-generic-ast-support/README.md

```diff
@@ -0,0 +1,62 @@
+- Repo: eslint/eslint
+- Start Date: 2020-06-04
+- RFC PR: 
+- Authors: Ilya Volodin
+
+# Generic AST support
+
+## Summary
+
+Allow ESLint to traverse and lint generic (non estree compliant AST) and allow parsers to specify information about AST that would drive ESLint's traversal and features.
+
+## Motivation
+
+ESLint has cemented itself as a go to tool for JavaScript/TypeScript ecosystem for style checking and linting. However, there are a lot of connected languages/technologies that are used every day by ESLint users that are either lacking linters, or require yet another tool to be setup and configured for the repository. Examples include HTML, CSS, SCSS, LESS, GraphQL, JSON, etc.
+With addition of config overrides users can now setup separate configuration per glob, so that would allow them to have a single instance of ESLint and a single configuration that would allow linting of the entire repository, including all non javascript/typescript files.
+
+## Detailed Design
+
+### Initial assumptions
+* This RFC does not expect or implies that existing core or plugin rules will work for any files that are parsed through non-estree compliant parsers. Every parser that chooses to output custom AST will have to provide new rules that will work for this AST in the form of plugins.
+* This RFC doesn't propose full feature parity for new parser with existing ESLint capabilities. For example, it doesn't propose a new way to provide custom Code Path Analyzer, or support inline directives for controlling ESLint's flow for custom ASTs. However, those functionalities can be added later with a separate RFC(s).
+
+Currently ESLint has a lot of hardcoded assumptions about AST. For example, it assumes that nodes are identified by the property called `type` and that property exists on every node. It assumes that top-most node is always `Program` and that AST supports code-path-analysis (which is very specific to JavaScript). This RFC proposes to modify expected output from the `parseForESLint` method of the custom parsers to include additional property called `parserSupportOptions` of type object with four new properties (for now):
+ - nodeIdentifier: string - Specifies name of the ASTNode property that identifies node. Defaults to `type` for backwards compatibility.
```

Comment:
I wonder if this should be a function instead of a string? Not all ASTs provide a string name for each node in a property. For example, the TypeScript AST has a `kind` property that is a number and then you need to look up the name from the number. Having a `getNodeType()` function instead of a `nodeIdentifier` string would allow us to support the TypeScript AST as well.

Reply:
Hmm.. Didn't know about that TypeScript quirk, sure. I can change it to a function.

Reply:
Updated to `function | undefined`.